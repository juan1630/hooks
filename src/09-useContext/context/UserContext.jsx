import { createContext } from 'react';



export const UserContext = createContext();
// esta funcio requiere un default value 