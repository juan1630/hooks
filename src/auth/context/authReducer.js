import { types } from "../types/types";

const intialState ={ logged: false };

export const authReducer = (state = intialState, action) =>{

    switch (action.type) {
        case types.login:
            return {
                
                logged: true,
                user: action.payload
            };
    
        case types.logout:
            return {
                ...state,
                logged: false,
            };
    
        default:
            state;
    }
};