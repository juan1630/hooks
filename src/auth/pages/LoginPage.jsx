import { Google } from "@mui/icons-material";
import { Button, Grid, Link, TextField, Typography } from "@mui/material";
import { Link as RouterLink } from 'react-router-dom'
import { AuthLayout } from "../layout/AuthLayout";
import {useForm} from "../../hooks/useForm.js";

import {checkingAutentication, startLoginCrendetials} from "../../store/auth/thunks.js";
import {useDispatch} from "react-redux";

export const LoginPage = () => {


    const dispatch = useDispatch();

    const { email, password, onInputChange, formState } = useForm({
        email: 'juan@gmail.com',
        password: 'juan123'
    });

    const onSubmit = ( event ) =>{
        event.preventDefault();
        dispatch(checkingAutentication());
    }


    const onGoogleSignIn = () => {
        console.log('On google sign in');

        dispatch(startLoginCrendetials());
    }

    return(
        <AuthLayout>
              <form onSubmit={onSubmit} >
                    <Grid container >
                        <Grid  item  xs={12} sx={{mt: 2}}  >
                            <TextField placeholder="email"  type='email' label="correo" fullWidth name='email'  onChange={onInputChange} value={ email } />
                        </Grid>

                        <Grid  item  xs={12} sx={{mt: 2}}  >
                            <TextField placeholder="password"  type='password' label="password" fullWidth name='password' onChange={onInputChange}  value={ password } />
                        </Grid>
                        <Grid container spacing={2} sx={{ mb:2, mt: 1 }}  >
                            <Grid item xs={ 12 }  sm={ 6 } >
                                <Button variant='contained' fullWidth  type='submit' > Log in</Button>
                            </Grid>
                            <Grid item xs={ 12 } sm={ 6 }  >
                                <Button variant='contained' fullWidth onClick={onGoogleSignIn} >
                                    <Google />
                                    <Typography sx={{ml:1}} >
                                        Log in

                                    </Typography>
                                </Button>
                            </Grid>
                        </Grid>
                        <Grid container direction='row' justifyContent='end' >
                            <Link color="inherit" to="/auth/register" component={ RouterLink }  >
                                Crear una cuenta 
                            </Link>
                        </Grid>
                    </Grid>
                </form>
        </AuthLayout>
    );
}