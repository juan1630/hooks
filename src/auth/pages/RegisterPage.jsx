import {Grid, TextField, Button, Typography, Alert} from "@mui/material";
import { Link } from "react-router-dom";
import { AuthLayout } from "../layout/AuthLayout";

import {  Link as RouterLink } from 'react-router-dom'
import {useForm} from "../../hooks/useForm";
import {useState, useMemo} from "react";
import {useDispatch, useSelector} from "react-redux";
import {startCreatingUserWithEmailPassword} from "../../store/auth/thunks.js";

const initial = {
    nombre:'juan',
    email:'juan@gmail.oom',
    password: '123456'
}

export const RegisterPage = ( ) => {


    const [ formSubmited, setFormSubmited ] = useState(false);
    const dispatch = useDispatch();

    const { status, errorMessage } = useSelector( state =>  state.auth );

    //memorizamos el valor de la depencia status, para no disparar el render
    const isCheckingAuthentication = useMemo( ()=> status === 'checking', [status] );

    const validationForms = {
        email: [ (value) => value.includes('@'), 'El correo debe ser valido'  ],
        password: [ (value) => value.length  >= 4, 'La contraseña debe de tener más de 6 letras' ],
        nombre: [ (value) => value.length >= 1 , 'El nombre es requerido' ]
    }

    const { onInputChange, nombre, email, password, formState,
            nombreValid, emailValid, passwordValid, isFormValid
        }  = useForm( initial, validationForms);

    const onSubmitInput = (e) => {
            e.preventDefault();
            setFormSubmited(true)

            if( !isFormValid ) return
            dispatch( startCreatingUserWithEmailPassword( formState ));
    }

    return(
    <AuthLayout >
        <form onSubmit={onSubmitInput}   className='animate__animated animate__fadeIn animate__faster' >
            <h1> Form Valid { (isFormValid) ? 'Valido' : 'Not valid ' } </h1>
              <Grid container >
                    <Grid  item  xs={12} sx={{mt: 2}}  >
                      <TextField placeholder="Juan Patrón" name='nombre'
                                 value={nombre}  onChange={onInputChange}
                                 type='text' label="Nombre" fullWidth
                                 error={ !!nombreValid && formSubmited }
                                 helperText={ nombreValid  }
                                />
                  </Grid>
                  <Grid  item  xs={12} sx={{mt: 2}}  >
                      <TextField placeholder="email" name='email'  value={email}  onChange={onInputChange}  type='email' label="correo" error={!!emailValid && formSubmited}  helperText={ emailValid }  fullWidth />
                  </Grid>

                  <Grid  item  xs={12} sx={{mt: 2}}  >
                      <TextField placeholder="password"  name='password'  value={password}  onChange={onInputChange} type='password' label="password"  error={ !!passwordValid && formSubmited }  helperText={passwordValid} fullWidth />
                  </Grid>
                  <Grid container spacing={2} sx={{ mb:2, mt: 1 }}  >
                      <Grid item display={ !!errorMessage ? '': 'none' } xs={12} >
                          <Alert severity='error'>
                              { !!errorMessage ? errorMessage.code : '' }
                          </Alert>
                      </Grid>
                      <Grid item xs={ 12 }  sm={ 6 } >
                          <Button
                                disabled={ isCheckingAuthentication }
                              type='submit'
                                variant='contained'
                                fullWidth>  Crear cuenta
                          </Button>
                      </Grid>
                  </Grid>
                  <Grid container direction='row' justifyContent='end' >
                        <Typography> ¿Ya tienes una cuenta ?  </Typography>
                      <Link color="inherit" to="/auth/login" component={ RouterLink }  >
                          Login 
                      </Link>
                  </Grid>
              </Grid>
          </form>
  </AuthLayout>
    );
}