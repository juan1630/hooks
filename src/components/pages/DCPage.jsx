import { HeroList } from "../heroes/HeroList";

export const DCPage = ()=>{
    return(
        <>
            <h1> DC page </h1>
            <hr />
            <HeroList publisher='DC Comics' />
        </>
    )
};