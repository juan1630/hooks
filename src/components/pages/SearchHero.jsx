import { useLocation, useNavigate } from "react-router-dom";
import { useForm } from "../../hooks/useForm";
import { HeroCard } from "../heroes/HeroCard";
import queryString from 'query-string';
import { getResultsByName } from "../helpers/getResultsByName";

export const SearchHeroe = ()=>{

    const navigate = useNavigate();
    const location = useLocation();

    const { q = '' } = queryString.parse(location.search);
    const heroes = getResultsByName(q);

    const showSearch = ( q.length === '' );
    const showError =  ( q.length > 0 && heroes.length == 0 )

    const {searchText, onInputChange}  = useForm({
        searchText:q
    });

    const onSearchSubmit = (event) =>{
        event.preventDefault();
        
        if(searchText.trim().length <= 1) return
        navigate('?q='+searchText.toLowerCase().trim());    
    };

    return(
        <>
            <h1> Search heroe </h1>
            <hr />
            <div className="row">
                <div className="col-5">
                    <h4> Searching... </h4>
                    <form  onSubmit={onSearchSubmit} >
                        <input type='text' 
                                className="form-control mt-4" 
                                placeholder="Search a hero" 
                                name="searchText" 
                                autoComplete="off"
                                onChange={onInputChange} 
                                value={searchText}
                                />

                                <button type="submit" className="btn btn-primary mt-4" > Search </button>
                    </form>
                </div>
                <div className="col-7 mt-4">

                    <h2>Results</h2>
                    
                    {/* {
                        (q=== '')
                        ? <div className="alert alert-primary">Search a hero</div>
                        : (heroes.length === 0) && <div className="alert alert-danger"><strong>Ther's no hero <b> { q } </b></strong></div>
                    } */}

                   
                        <div className="alert alert-primary animate__animated animate__fadeIn"  style={{ display:  (showSearch) ? '' : 'none' }} >Search a hero</div>
                        <div className="alert alert-danger animate__animated animate__fadeIn" style={{ display: (showError) ? '': 'none' }} ><strong>Ther's no hero <b> { q } </b></strong></div> 
            
                    
                        {
                            heroes.map( hero => <HeroCard key={hero.id} {...hero} /> )
                        }
                </div>
            </div>
        </>
    );
};