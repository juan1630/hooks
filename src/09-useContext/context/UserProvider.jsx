import { useState } from "react";
import { UserContext } from "./UserContext";

 export const UserProvider = ({ children }) =>{ 

    const [user, setUser] = useState();

    return( 
        // hacemos visible este value para todos los componentes hijos que se encuentren dentro del provider 
        // <UserContext.Provider  value={{hola: 'mundo', user}} >
        <UserContext.Provider value={{
            user,
            setUser
        }}>
            {children}
        </UserContext.Provider>
    );
 };