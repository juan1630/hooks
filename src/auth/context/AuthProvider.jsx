import { useReducer } from 'react';
import { types } from '../types/types';
import { AuthContext } from './AuthContext';
import { authReducer } from './authReducer';


const initialState ={
    logged: false
}


const init = ()=> {
    const user = JSON.parse(localStorage.getItem('user'));

    return {
        logged: !!user,
        user
    }
}


export const AuthProvider = ({children}) =>{

    const [ authState, dispatch ] = useReducer(authReducer, initialState, init);

    const login = ( name = '')=>{
        const action ={ 
            type: types.login,
            payload: name
        }

        localStorage.setItem('user', name);

        dispatch( action );
    }

    const logout = () => {

        const action ={
            type: types.logout
        }
        
        localStorage.removeItem('user');
        dispatch( action );
    }
// hacemos publicos los paramentros ya que se inyectan en el componente 
    return(
        <AuthContext.Provider value={{
            ...authState,
            //methods
            login,
            logout
        }} >
            { children }
        </AuthContext.Provider>
    );
}