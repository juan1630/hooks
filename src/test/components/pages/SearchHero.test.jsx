import { fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { SearchHeroe } from "../../../components/pages/SearchHero";

import { AuthContext } from "../../../auth/context/AuthContext";

const mockedUseNavigate = jest.fn();

jest.mock('react-router-dom', ()=>({
    ...jest.requireActual('react-router-dom'),
    useNavigate: ()=> mockedUseNavigate
}));


describe('Pruebas en el componente seacrhHero', ()=> {



    beforeEach( ()=> jest.clearAllMocks());


    test('Debe de mostrarse correctamente con los valores por defecto', ()=>{
        const container = render(
            <MemoryRouter>

                    <SearchHeroe />
            </MemoryRouter>
        );

            // para actulizar el snapshot se usa la U
        expect( container ).toMatchSnapshot();
        // screen.debug();
    });


    test('Debe de mostrar a Batman y el valor del queryString', ()=>{
        render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                {/* pasamos al url los query parameters, en este caso el de batman */}
                <SearchHeroe />
            </MemoryRouter>
        )

            const inputValue = screen.getByRole('textbox');

            expect( inputValue.value ).toBe('batman')
            // compatramso el input con el valor de batman

            const imgValue = screen.getByRole('img');

            expect( imgValue.src ).toBe('http://localhost/assets/heroes/dc-batman.jpg');
    
            // screen.debug();

            const alertHero = screen.getByLabelText('alertNoHero').style._values;
            
            // expect( alertHero ).toEqual( { display: 'none' })
            expect(alertHero.display).toBe('none')
        }); 


        test('Deba de mostrar el alert sino encuntra a batman123', ()=>{
            render(
                <MemoryRouter initialEntries={['/search?q=batman123']}>
                    {/* pasamos al url los query parameters, en este caso el de batman */}
                    <SearchHeroe />
                </MemoryRouter>
            );

            const alertHero = screen.getByLabelText('alertNoHero').style;
            
            // expect( alertHero ).toEqual( { display: 'none' })
            expect(alertHero.display).toBe('')

        });


        test('debe de mandar a llamar el navigate', ()=>{
            
            const valueContext = {
                logged: true
            }
            
            render(
                <AuthContext.Provider value={ valueContext } >
                    <MemoryRouter   initialEntries={['/search']} >
                        <SearchHeroe />
                    </MemoryRouter>
                </AuthContext.Provider>
            )
                const inptuSearche = screen.getByRole('textbox');

                fireEvent.change(inptuSearche, { target:{ name: 'searchText', value: 'superman' }});



                const form = screen.getByRole('form');

                // console.log(form);

                fireEvent.submit(form);




                // console.log( inptuSearche.value )

                expect( mockedUseNavigate ).toHaveBeenCalledWith('?q=superman')
        });

});