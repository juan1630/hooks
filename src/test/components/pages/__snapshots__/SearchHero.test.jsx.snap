// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente seacrhHero Debe de mostrarse correctamente con los valores por defecto 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <h1>
         Search heroe 
      </h1>
      <hr />
      <div
        class="row"
      >
        <div
          class="col-5"
        >
          <h4>
             Searching... 
          </h4>
          <form
            aria-label="form"
          >
            <input
              autocomplete="off"
              class="form-control mt-4"
              name="searchText"
              placeholder="Search a hero"
              type="text"
              value=""
            />
            <button
              class="btn btn-primary mt-4"
              type="submit"
            >
               Search 
            </button>
          </form>
        </div>
        <div
          class="col-7 mt-4"
        >
          <h2>
            Results
          </h2>
          <div
            class="alert alert-primary animate__animated animate__fadeIn"
            style="display: none;"
          >
            Search a hero
          </div>
          <div
            aria-label="alertNoHero"
            class="alert alert-danger animate__animated animate__fadeIn"
            style="display: none;"
          >
            <strong>
              Ther's no hero 
              <b>
                 
                 
              </b>
            </strong>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <h1>
       Search heroe 
    </h1>
    <hr />
    <div
      class="row"
    >
      <div
        class="col-5"
      >
        <h4>
           Searching... 
        </h4>
        <form
          aria-label="form"
        >
          <input
            autocomplete="off"
            class="form-control mt-4"
            name="searchText"
            placeholder="Search a hero"
            type="text"
            value=""
          />
          <button
            class="btn btn-primary mt-4"
            type="submit"
          >
             Search 
          </button>
        </form>
      </div>
      <div
        class="col-7 mt-4"
      >
        <h2>
          Results
        </h2>
        <div
          class="alert alert-primary animate__animated animate__fadeIn"
          style="display: none;"
        >
          Search a hero
        </div>
        <div
          aria-label="alertNoHero"
          class="alert alert-danger animate__animated animate__fadeIn"
          style="display: none;"
        >
          <strong>
            Ther's no hero 
            <b>
               
               
            </b>
          </strong>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
